retrieval/scrape.go

```diff
@@ -0,0 +1,292 @@
+	if sl.active() {
+		return
+	}
+
+	// If no offset function is provided, default to starting immediately.
+	if sl.offset == nil {
+		sl.offset = func(time.Duration) time.Duration { return 0 }
+	}
+
+	ctx, sl.cancel = context.WithCancel(ctx)
+
+	sl.stopc = make(chan struct{})
+	sl.done = make(chan struct{})
+
+	defer func() {
+		close(sl.done)
+		sl.cancel = nil
+	}()
+
+	select {
+	case <-time.After(sl.offset(interval)):
+		// Continue after a scraping offset.
+	case <-ctx.Done():
+		return
+	case <-sl.stopc:
+		return
+	}
+
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	for {
+		start := time.Now()
+
+		err := sl.scrape(ctx, timeout)
+		if err != nil {
+			log.Debugf("Scraping error: %s", err)
+		}
+
+		sl.report(start, time.Since(start), err)
+		if errc != nil {
+			errc <- err
+		}
+
+		// First check without the ticker to not initiate
+		// another scrape if termination was requested.
+		select {
+		case <-sl.stopc:
+			return
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		select {
+		case <-sl.stopc:
+			return
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+		}
+	}
+}
+
+// kill terminates the scraping loop immediately.
+func (sl *scrapeLoop) kill() {
+	if !sl.active() {
+		return
+	}
+	sl.cancel()
+}
+
+// stop finishes any pending scrapes and then kills the loop.
+func (sl *scrapeLoop) stop() {
+	if !sl.active() {
+		return
+	}
+	close(sl.stopc)
+	<-sl.done
+
+	sl.kill()
+}
+
+// scrape executes a single Scrape and appends the retrieved samples.
+func (sl *scrapeLoop) scrape(ctx context.Context, timeout time.Duration) error {
+	ch := make(chan *model.Sample, ingestedSamplesCap)
+
+	// Receive and append all the samples retrieved by the scraper.
+	go func() {
+		var (
+			smpl *model.Sample
+			ok   bool
+		)
+		for {
+			select {
+			case smpl, ok = <-ch:
+				if !ok {
+					return
+				}
+			case <-ctx.Done():
+				return
+			}
+			// TODO(fabxc): Change the SampleAppender interface to return an error
+			// so we can proceed based on the status and don't leak goroutines trying
+			// to append a single sample after dropping all the other ones.
+			sl.appender.Append(smpl)
+		}
+	}()
+
+	ctx, _ = context.WithTimeout(ctx, timeout)
+
+	return sl.scraper.Scrape(ctx, ch)
+}
+
+func (sl *scrapeLoop) report(start time.Time, duration time.Duration, err error) {
+	sl.state.setLastScrape(start)
+	sl.state.setLastError(err)
+
+	ts := model.TimeFromUnixNano(start.UnixNano())
+
+	var health model.SampleValue
+	if err == nil {
+		health = 1
+	}
+
+	healthSample := &model.Sample{
+		Metric: model.Metric{
+			model.MetricNameLabel: scrapeHealthMetricName,
+		},
+		Timestamp: ts,
+		Value:     health,
+	}
+	durationSample := &model.Sample{
+		Metric: model.Metric{
+			model.MetricNameLabel: scrapeDurationMetricName,
+		},
+		Timestamp: ts,
+		Value:     model.SampleValue(float64(duration) / float64(time.Second)),
+	}
+
+	sl.appender.Append(healthSample)
```

Comment:
Metric relabelling shouldn't apply to the non-scraped samples. I thought I'd a unittest for that.